kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,612 @@
+# KEP - Real Load Aware Scheduling
+
+
+## Summary
+
+Minimizing machine costs by utilizing all nodes is the main objective for efficient cluster management. To achieve this goal, we can make the Kubernetes scheduler aware of the gap between resource allocation and actual resource utilization. Taking advantage of the gap may help pack pods more efficiently, while the default scheduler that only considers pod requests and allocable resources on nodes cannot.
+
+## Motivation
+
+Kubernetes provides a declarative resource model that core components (scheduler and kubelet) honor to behave consistently and satisfy QoS guarantees.
+However, using this model can lead to low-utilization of the cluster for the following reasons:
+
+1. It is hard for users to estimate accurate resource usage for applications. Also, users may not understand the resource model and not set it at all.
+2. The default in-tree scheduling plugins (Score) that Kubernetes provides don't consider live node utilization values.
+
+This proposal utilizes real time resource usage to schedule pods with proposed plugins. The eventual goal is to increase cluster utilization and decrease costs of cluster management without breaking the contract of Kubernetes Resource Model.
+
+### Goals
+
+1. Provide configurable scheduling plugins to increase cluster utilization.
+2. Expected CPU Utilisation per node should not go beyond X%.
+3. Instances of an app should spread across failure domains. The failure domain would be host at the lowest level and can extend to racks, zones, regions, etc.
+4. To not affect the behavior of default score plugins unless necessary.
+5. Implement the above features as Score plugins
+
+### Non-Goals
+
+1. Implement the above constraints (2, 3) as Filter plugins.
+2. Descheduling due to unexpected outcome (hot nodes, fragmentation etc.) of past scoring by plugins is not addressed in the initial design.
+3. Memory, Network, and Disk utilization are not considered in the initial design.
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+
+As a company relying on the public cloud, we would like to minimize machine costs by efficiently utilizing the nodes leased.
+
+
+As a company-owned data center, we would like to minimize the hardware and maintenance associated with clusters by getting the maximum possible juice out of existing machines.
+
+
+#### Story 2
+
+Increasing resource utilization as much as possible may not be the right solution for all clusters. As it always takes some time to scale up a cluster to handle the sudden spikes of load, cluster admins would like to leave adequate room for the bursty load, so there is enough time to add more nodes to the cluster.
+
+
+### Notes/Constraints/Caveats
+
+Enabling our plugin(s) will cause conflict with 2 default scoring plugins: "NodeResourcesLeastAllocated" and "NodeResourcesBalancedAllocation" plugins. So it is strongly advised to disable them when enabling plugin(s) mentioned in this proposal.
```

Comment:
Will this also conflict with `NodeResourcesMostAllocated`? (it's disabled by default but good to mention if so). The bin packing plugin seems similar to it

Reply:
Yes it will. I added that.

Reply:
Can you split the conflicts for each of your plugins? Do they both conflict with all the plugins mentioned here?

Reply:
They both conflict with plugins mentioned, hence the `plugin(s)` part.